pkg/applicationmanager/controllers/applicationbackup.go

```diff
@@ -469,225 +469,237 @@ func (a *ApplicationBackupController) backupVolumes(backup *stork_api.Applicatio
 	// Start backup of the volumes if we don't have any status stored
 	pvcMappings := make(map[string][]v1.PersistentVolumeClaim)
 	backup.Status.Stage = stork_api.ApplicationBackupStageVolumes
-
-	objectMap := stork_api.CreateObjectsMap(backup.Spec.IncludeResources)
-	info := stork_api.ObjectInfo{
-		GroupVersionKind: metav1.GroupVersionKind{
-			Group:   "core",
-			Version: "v1",
-			Kind:    "PersistentVolumeClaim",
-		},
-	}
-
-	var pvcCount int
-	for _, namespace := range backup.Spec.Namespaces {
-		pvcList, err := core.Instance().GetPersistentVolumeClaims(namespace, backup.Spec.Selectors)
-		if err != nil {
-			return fmt.Errorf("error getting list of volumes to migrate: %v", err)
+	namespacedName := types.NamespacedName{}
+	if IsVolsToBeBackedUp(backup) {
+		isResourceTypePVC := IsResourceTypePVC(backup)
+		objectMap := stork_api.CreateObjectsMap(backup.Spec.IncludeResources)
+		info := stork_api.ObjectInfo{
+			GroupVersionKind: metav1.GroupVersionKind{
+				Group:   "core",
+				Version: "v1",
+				Kind:    "PersistentVolumeClaim",
+			},
 		}
 
-		for _, pvc := range pvcList.Items {
-			// If a list of resources was specified during backup check if
-			// this PVC was included
-			info.Name = pvc.Name
-			info.Namespace = pvc.Namespace
-			if len(objectMap) != 0 {
-				if val, present := objectMap[info]; !present || !val {
-					continue
-				}
+		var pvcCount int
+		for _, namespace := range backup.Spec.Namespaces {
+			pvcList, err := core.Instance().GetPersistentVolumeClaims(namespace, backup.Spec.Selectors)
+			if err != nil {
+				return fmt.Errorf("error getting list of volumes to migrate: %v", err)
 			}
 
-			// Don't backup pending or deleting PVCs
-			if pvc.Status.Phase != v1.ClaimBound || pvc.DeletionTimestamp != nil {
-				continue
-			}
-			driverName, err := volume.GetPVCDriver(core.Instance(), &pvc)
-			if err != nil {
-				// Skip unsupported PVCs
-				if _, ok := err.(*errors.ErrNotSupported); ok {
+			for _, pvc := range pvcList.Items {
+				// If a list of resources was specified during backup check if
+				// this PVC was included
+				info.Name = pvc.Name
+				info.Namespace = pvc.Namespace
+				if len(objectMap) != 0 {
+					if resourcecollector.IsNsPresentInIncludeResource(objectMap, namespace) {
+						if val, present := objectMap[info]; !present || !val {
+							continue
+						}
+					} else {
+						// We could have case where includeResource has data, current ns is not part of includeResource
+						// and the user has given ResourceType != PVC. In this case we don't
+						// want to collect vol data from this ns
+						if !isResourceTypePVC {
```

Comment:
This check could be outside the PVC for loop as it is generic. It need not be run for each PVC items.

Reply:
What I meant is that, we can move the check of moving to next NS outside. This need not to be done for PVC and   
then move to the next NS. Let say the line number 506 continue hits for every PVC that condition can be moved out.
```
if len(objectMap) != 0 {					
        if !resourcecollector.IsNsPresentInIncludeResource(objectMap, namespace) {
                   
               if !isResourceTypePVC {
               		continue
               }
        }
 }

Reply:
Lets take the following CRD 
```apiVersion: stork.libopenstorage.org/v1alpha1
kind: ApplicationBackup
metadata:
  name: case-9
  namespace: kube-system
spec:
  backupLocation: mysql
  namespaces:
  - ns1
  - ns2
  - ns3
  resourceTypes:
  - ConfigMap
  includeResources:
  - name: ns2-pvc-2
    group: core
    kind: PersistentVolumeClaim
    namespace: ns2
    version: v1
  reclaimPolicy: Delete
  selectors:
  preExecRule:
  postExecRule:
```

 In the above CRD, we would like to backup the following
 - Only PVC from ns2
 - For ns1 and ns3, we want to skip vol backups and only backup resouces.

This check if for every ns to know what they need to be backed up.
If a ns is not part of the includeResource list but includeResource has refererence to non-pvc resource, then we need to be skipping those resoruce.
Moving this to before ns check would not solve this issue and we would end up backing vols data for those ns even though ResourceType != PVC.

This check has to be executed only when the ns is not part of includeResource list and doesn't have any PVC  hence added here.

Reply:
As discussed, the check "if !isResourceTypePVC " in 505 line, will done for all the PVC in the namespace even if the namespaces is not part of includeResources and global ResourceType does not have PVC in it. 
When we decide that we need to skip to next namespace in line 505, either we can have break instead of continue, that way we can skip pvc for loop Or have check outside the PVC for loop.

Reply:
Addressed and commited.